src/Tools/DependencyManagement/ReflectionContainer.php

```diff
@@ -0,0 +1,49 @@
+<?php
+
+namespace Automattic\WooCommerce\Tools\DependencyManagement;
+
+/**
+ * The original PHP League's `ReflectionContainer`, used for autowiring, provides an all-or-nothing approach
```

Comment:
Sure!

The default container class has a `share` method to declare shared instances, but that class one doesn't have autowiring, meaning that you need to register all the classes you intend to resolve AND also register all the constructor arguments for those.

To ease this there's the `ReflectionContainer`, you don't need to register anything and it will resolve any class and their arguments based on type hints, which is very convenient. However this one doesn't have a `share` method: either no instance is shared (default), or all are (if set via `cacheResolutions` method).

So my class gives us the best of both worlds: you can define only a given set of classes to be shared, while still keeping the benefit of autowiring.

Reply:
By the looks of the [autowiring documentation](https://container.thephpleague.com/3.x/auto-wiring/), this may just be caused by trying to use `get` on the ReflectionContainer itself. I don't think we actually need to hold onto a reference to the container and can just call `share` on the `Container` to register a singleton.

Reply:
I've looked a bit more and I think I understand a little bit better now, sorry for the misunderstanding! 

The `ReflectionContainer` has a `cacheResolutions` property that causes it to turn every instantiated class into a shared instance if set. The normal `Container` works by attempting to resolve an instance each time the container is interacted with. If it fails to find the instance in the definitions, inflectors, and service providers, then it uses the delegate container to perform the resolution. The `ReflectionContainer` is then accessed at this point and resolves all of the constructor parameters out of the original `Container` instance. It'll even do this in a nested manner and auto-wire deeply if desired!

The only thing you lose calling `get()` on the `Container` vs the `ReflectionContainer` is the ability to override constructor parameters with an array, but I think in general that tests shouldn't be resolving things using the container anyway right? With that said too we might consider some mechanism for manipulating the container directly in tests that isn't possible in normal code, but I'm not sure what we might do to achieve that!

Reply:
I'm afraid that you still don't get it 🙂 I mean that we are talking about different things, not that you don't understand how the container works.

Ok, so the regular container allows you to register explicit instances of classes with `add(class_name, the_instance)`. That's great for specific cases, but in general what we want is the container to be able to create new instances by itself whenever needed.

And this is of course possible, by using just `add(class_name)`. But having to do that for each and every class we intend to use is cumbersome. Moreover, when using this approach we need to declare all the constructor parameters, so if we have `__contruct(A a, B b, C c)` we need to do `add(class_name)->addArgument(A:class)->addArgument(B:class)->addArgument(C:class)` - extra cumbersomeness.

Now, regarding shared instances, we can declare a given class as shared by using the `share` method. However that method is just "syntactic sugar" for `add()->setShared()`, so we still need to declare the constructor arguments.

As for autowiring: adding a fallback `ReflectionContainer` saves us from having to declare all the classes we want to use, with the added bonus that it works regardless of any constructor arguments in place. BUT out of the box that container doesn't allow to declare individual classes as shared - either you use `cacheResolutions` and then all the resolved instances are shared, or you don't and none is.

That's where my custom `ReflectionContainer` comes into play. It has a `shared` method that is similar to the method with the same name in the regular container, and in the end server the same purpose - with the difference that you don't need to specify any constructor argument.

So, when using `share(class_name)` in my `ReflectionContainer` you are telling it "Hey, whenever someone requests a class_name, use reflection to autowire everything, and once you have the instance cache it and always return the same one".

Reply:
Discussed in Slack